operators/src/pro/cache/tile_cache.rs

```diff
@@ -29,16 +30,425 @@ pub struct TileCacheBackend {
+        }
+
+        Ok(query_id)
+    }
+
+    fn remove_query_from_landing_zone(&mut self, key: &CanonicOperatorName, query_id: &QueryId) {
+        if let Some((landing_zone, landing_zone_size)) = self.landing_zone_and_size_mut(key) {
+            if let Some(entry) = landing_zone.remove(query_id) {
+                landing_zone_size.remove_element_bytes(query_id);
+                landing_zone_size.remove_element_bytes(&entry);
+            }
+        } else {
+            debug!("Tried to remove query ({}) from landing zone, but there was no entry for the operator ({:?})!", query_id, key);
+        }
+    }
+
+    /// Insert a tile for a given query. The query has to be inserted first.
+    /// The tile is inserted into the landing zone and only moved to the cache when the query is finished.
+    /// If the landing zone is full, the caching of the query is aborted.
+    pub fn insert_tile<T>(
+        &mut self,
+        key: &CanonicOperatorName,
+        query_id: &QueryId,
+        tile: RasterTile2D<T>,
+    ) -> Result<(), CacheError>
+    where
+        T: Pixel + Cachable,
+    {
+        // if the tile is already expired we can stop caching the whole query
+        if tile.cache_hint.is_expired() {
+            // TODO: cache hint as trait
+            self.remove_query_from_landing_zone(key, query_id);
+            debug!(
+                "Tile expired before insertion {}. Is empty: {}",
+                tile.cache_hint.expires().datetime(),
+                tile.is_empty()
+            );
+            return Err(super::error::CacheError::TileExpiredBeforeInsertion);
+        }
+
+        let tile_size_bytes = tile.byte_size();
+
+        // check if landing zone has enough space, otherwise abort caching the query
+        if !self.landing_zone_size.can_fit_bytes(tile_size_bytes) {
+            debug!(
+                "Not enough space in landing zone. Removing query {}. Landing zone size: {}. Landing zone used: {}. Landing zone used percentage: {}. Tile size: {}",
+                query_id, self.landing_zone_size.total_byte_size(), self.landing_zone_size.byte_size_used(), self.landing_zone_size.size_used_fraction(), tile_size_bytes
+            );
+            self.remove_query_from_landing_zone(key, query_id);
+            return Err(super::error::CacheError::NotEnoughSpaceInLandingZone);
+        }
+
+        let (landing_zone, landing_zone_size) = self
+            .landing_zone_and_size_mut(key)
+            .ok_or(super::error::CacheError::QueryNotFoundInLandingZone)?;
+
+        let entry = landing_zone
+            .get_mut(query_id)
+            .ok_or(super::error::CacheError::QueryNotFoundInLandingZone)?;
+
+        entry.query.spatial_bounds.extend(&tile.spatial_partition()); // since the source should only produce tiles that intersect with the query, we can extend the query bounds
```

Comment:
can you maybe improve the comment here a bit? Like, that we are not using the spatial_bounds of the query but of the result, which may be larger then the original query because we always compute full tiles?

Reply:
587fb99972ca447a9052573250432a5023946202